substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jfr/JfrEventSubstitution.java

```diff
@@ -66,14 +67,20 @@ public class JfrEventSubstitution extends SubstitutionProcessor {
     private final ConcurrentHashMap<ResolvedJavaField, ResolvedJavaField> fieldSubstitutions;
     private final Map<String, Class<? extends jdk.jfr.Event>> mirrorEventMapping;
 
+    private static final Method registerMirror = JavaVersionUtil.JAVA_SPEC < 22 ? ReflectionUtil.lookupMethod(SecuritySupport.class, "registerMirror", Class.class) : null;
+
     JfrEventSubstitution(MetaAccessProvider metaAccess) {
         baseEventType = metaAccess.lookupJavaType(jdk.internal.event.Event.class);
         ResolvedJavaType jdkJfrEventWriter = metaAccess.lookupJavaType(EventWriter.class);
         changeWriterResetMethod(jdkJfrEventWriter);
         typeSubstitution = new ConcurrentHashMap<>();
         methodSubstitutions = new ConcurrentHashMap<>();
         fieldSubstitutions = new ConcurrentHashMap<>();
-        mirrorEventMapping = createMirrorEventsMapping();
+        if (JavaVersionUtil.JAVA_SPEC < 22) {
+            mirrorEventMapping = createMirrorEventsMapping();
+        } else {
+            mirrorEventMapping = null;
```

Comment:
For JDK 22, I think you will still need to call `createMirrorEventsMapping()`. But `createMirrorEventsMapping()` will need `ReflectionUtil.lookupClass(false, "jdk.jfr.internal.MirrorEvents");` instead of `ReflectionUtil.lookupClass(false, "jdk.jfr.internal.instrument.JDKEvents");` depending on what `JavaVersionUtil.JAVA_SPEC` returns. 

Additionally, I think an `@Alias` will be required for `jdk.jfr.internal.instrument.JDKEvents#mirrorEventClasses` similar to what already exists in `Target_jdk_jfr_internal_instrument_JDKEvents` for JDK 21.

Reply:
Thanks so much for the suggestions @roberttoyonaga! Interestingly, the mirrored events are not covered by our testing.

Since the JDK update is rather critical, we plan to merge this PR as is and deal with the remaining JFR changes in a follow-up PR. We should also add some testing so that we cover these cases in our gate jobs.

If you could prepare a PR, we would very much appreciate it as it would speed up things. I can let you know as soon as the labjdk builds are publicly available, so you can start on top of this PR without waiting for it to merge.

Otherwise, if you don't have time, I can prepare a PR as well.

Reply:
The labsJDK 22+22 artifacts are available: https://github.com/graalvm/labs-openjdk/releases

Reply:
> If you could prepare a PR, we would very much appreciate it as it would speed up things.

Hi @zapster! Yes, I'll prepare a PR for these JFR updates

Reply:
Hi @zapster, on second thought, we may not actually need to do anything further to accommodate the changes related to `MirrorEvents.java`.  The new behaviour is that `jdk.jfr.internal.MirrorEvents` maintains a cache of all the mirror event classes instead of that cache needing to be built in `JDKEvents`. So this means the extra step of using `SecuritySupport#registerMirror` is unnecessary. We also now don't have to worry about doing this extra step at native image build-time before manually retransforming the JFR event classes in `JfrEventSubstitution` anymore.   

I made a PR that adds a test to verify mirror events are handled correctly. https://github.com/oracle/graal/pull/7771

Reply:
Awesome, thanks a lot!